src/core/algorithms/cfd/cfd_discovery.cpp

```diff
@@ -50,11 +50,36 @@ int CFDDiscovery::NrCfds() const {
     return (int)cfd_list_.size();
 }
 
-CFDList CFDDiscovery::GetCfds() const {
+ItemsetCFDList const& CFDDiscovery::GetItemsetCfds() const {
     return cfd_list_;
 }
 
-std::string CFDDiscovery::GetCfdString(CFD const& cfd) const {
+CFDList CFDDiscovery::GetCfds() const {
+    CFDList list;
+
+    auto toCFD = [&rel = *relation_](ItemsetCFD const& dep) -> CFD {
+        auto& [lhs, rhs] = dep;
+
+        CFD::RawItems cfd_lhs;
+        for (uint i = 0; i < lhs.size(); i++) {
+            Item item = lhs[i];
+            int attr = item == 0 ? (int)i : Output::ItemToAttrIndex(item, rel);
+            std::optional<std::string> pattern_opt = Output::ItemToPatternOpt(item, rel);
+            cfd_lhs.push_back({.attribute = attr, .value = Output::ItemToPatternOpt(item, rel)});
+        }
+
+        CFD::RawItem cfd_rhs = {.attribute = Output::ItemToAttrIndex(rhs, rel),
+                                .value = Output::ItemToPatternOpt(rhs, rel)};
+        return CFD{cfd_lhs, cfd_rhs};
+    };
+
+    for (ItemsetCFD const& dep : GetItemsetCfds()) {
+        list.emplace_back(toCFD(dep));
```

Comment:
nit: `push_back` since we don't construct value in-place from constructor arguments here actually